craft_application/git/_git_repo.py

```diff
@@ -0,0 +1,310 @@
+        """
+        try:
+            # for a clean repo, `status()` will return an empty dict
+            return not bool(self._repo.status())
+        except pygit2.GitError as error:
+            raise GitError(
+                f"Could not check if the git repository in {str(self.path)!r} is clean."
+            ) from error
+
+    def _init_repo(self) -> None:
+        """Initialize a git repo.
+
+        :raises GitError: if the repo cannot be initialized
+        """
+        logger.debug("Initializing git repository in %r", str(self.path))
+
+        try:
+            pygit2.init_repository(  # pyright: ignore[reportUnknownMemberType]
+                self.path
+            )
+        except pygit2.GitError as error:
+            raise GitError(
+                f"Could not initialize a git repository in {str(self.path)!r}."
+            ) from error
+
+    def push_url(  # noqa: PLR0912 (too-many-branches)
+        self,
+        remote_url: str,
+        remote_branch: str,
+        ref: str = "HEAD",
+        token: str | None = None,
+        *,
+        push_tags: bool = False,
+    ) -> None:
+        """Push a reference to a branch on a remote url.
+
+        :param remote_url: the remote repo URL to push to
+        :param remote_branch: the branch on the remote to push to
+        :param ref: name of shorthand ref to push (i.e. a branch, tag, or `HEAD`)
+        :param token: token in the url to hide in logs and errors
+        :param push_tags: if true, push all tags to URL (similar to `git push --tags`)
+
+        :raises GitError: if the ref cannot be resolved or pushed
+        """
+        resolved_ref = self._resolve_ref(ref)
+        refspec = f"{resolved_ref}:refs/heads/{remote_branch}"
+
+        # hide secret tokens embedded in a url
+        stripped_url = remote_url.replace(token, "<token>") if token else remote_url
+
+        logger.debug(
+            "Pushing %r to remote %r with refspec %r.", ref, stripped_url, refspec
+        )
+
+        # temporarily call git directly due to libgit2 bug that unable to push
+        # large repos using https. See https://github.com/libgit2/libgit2/issues/6385
+        # and https://github.com/snapcore/snapcraft/issues/4478
+        # Force push in case this repository already exists. The repository is always
+        # going to exist solely for remote builds, so the only potential issue here is a
+        # race condition with multiple remote builds on the same machine.
+        cmd: list[str] = ["git", "push", "--force", remote_url, refspec, "--progress"]
+        if push_tags:
+            cmd.append("--tags")
+
+        git_proc: subprocess.Popen[str] | None = None
+        try:
+            with subprocess.Popen(
+                cmd,
+                cwd=str(self.path),
+                bufsize=1,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                universal_newlines=True,
+            ) as git_proc:
```

Comment:
Maybe add a comment explaining why you need to use Popen rather than subprocess.run?